### PR TITLE
Extend 'describe' with support for more messages

### PR DESCRIPTION
**Pull request by yallop**
Extend 'describe' with support for more messages and tidy up error, string and file handling in the `fusedump` code.

### CODE REVIEW

File: lib/profuse_7_23.ml [hunk 1]
```diff
@@ -873,18 +892,34 @@ module In = struct
...
+         | Setxattr r ->
+           let size = getf r Setxattr.T.size in
+           let flags = getf r Setxattr.T.flags in
+           Printf.sprintf "size=%ld flags%ld="
+             (UInt32.to_int32 size)
+             (UInt32.to_int32 flags)
+         | Listxattr r ->
+           let size = getf r Getxattr.T.size in
+           Printf.sprintf "size=%ld"
+             (UInt32.to_int32 size)
```
Comment (dsheets) [inline lib/profuse_7_23.ml:922]: `Getxattr` and `Setxattr` are missing the name strings.

File: lib/profuse_7_23.ml [hunk 2]
```diff
@@ -873,18 +892,34 @@ module In = struct
...
+         | Removexattr name -> name
+         | Getlk t ->
+           Lk.describe t
+         | Setlk t ->
+           Lk.describe t
+         | Setlkw t ->
+           Lk.describe t
+         | Link (l,n) ->
+           let name = Link.name (to_voidp (addr l)) in
+           Printf.sprintf "name=%s %s" name n
```
Comment (dsheets) [inline lib/profuse_7_23.ml]: `Link` prints the name twice (`name` was already parsed and put in the constructor) and does not print the oldnodeid from the structure.

File: lib/profuse_7_23.ml [hunk 3]
```diff
@@ -873,18 +892,34 @@ module In = struct
...
+         | Flush _r -> ""
+         | Fsyncdir _r -> ""
+         | Fsync _r -> ""
+         | Bmap _r -> ""
```
Comment (dsheets) [inline lib/profuse_7_23.ml]: I would rather `Flush`, `Fsyncdir`, `Fsync`, and `Bmap` say FIXME or TODO or something so the user knows that the serialization is not complete.

**Comment by dsheets** [comment]
I haven't tested the refactoring parts but if it works then I'm very happy with it. Thanks for cleaning that stuff up!

File: lib/profuse_7_23.ml (hunk 2 above)
Comment (yallop) [reply lib/profuse_7_23.ml]: Fixed (a00028e).

File: lib/profuse_7_23.ml (hunk 3 above)
Comment (yallop) [reply lib/profuse_7_23.ml]: Fixed (0c766e6).

File: lib/profuse_7_23.ml [hunk 4]
```diff
@@ -379,6 +390,14 @@ module In = struct
...
   module Lk = struct
     module T = T.Lk
+
+    let describe lk =
+      let fh = getf lk T.fh
+      and owner = getf lk T.owner
+      and lk = getf lk T.lk in
+      Printf.sprintf "fh=%Ld owner=%Ld lk=(%s)"
+        fh (UInt64.to_int64 owner)
+        (Struct.File_lock.describe lk)
```
Comment (dsheets) [inline lib/profuse_7_23.ml:401]: I think this is missing `lk_flags`?

File: lib/profuse_7_23.ml [hunk 5]
```diff
@@ -873,18 +902,37 @@ module In = struct
...
-         | Bmap _ -> "FIX ME"
+         | Getxattr (r, name) ->
+           let size = getf r Getxattr.T.size in
+           Printf.sprintf "name=%s size=%ld"
+             name
+             (UInt32.to_int32 size)
+         | Setxattr (r, name) ->
+           let size = getf r Setxattr.T.size in
+           let flags = getf r Setxattr.T.flags in
+           Printf.sprintf "name=%s size=%ld flags%ld="
```
Comment (dsheets) [inline lib/profuse_7_23.ml]: s/flags%ld=/flags=%ld/

File: lib/profuse_7_8.ml [hunk 6]
```diff
@@ -792,18 +811,34 @@ module In = struct
...
-         | Fsync _
-         | Bmap _ -> "FIX ME"
+         | Getxattr r ->
+           let size = getf r Getxattr.T.size in
+           Printf.sprintf "size=%ld"
+             (UInt32.to_int32 size)
+         | Setxattr r ->
+           let size = getf r Setxattr.T.size in
+           let flags = getf r Setxattr.T.flags in
+           Printf.sprintf "size=%ld flags%ld="
```
Comment (dsheets) [inline lib/profuse_7_8.ml]: s/flags%ld=/flags=%ld/

File: lib/profuse_7_8.ml [hunk 7]
```diff
@@ -792,18 +811,34 @@ module In = struct
...
-         | Removexattr _
-         | Getlk _
-         | Setlk _
-         | Setlkw _
-         | Link (_,_)
-         | Flush _
-         | Fsyncdir _
-         | Fsync _
-         | Bmap _ -> "FIX ME"
+         | Getxattr r ->
```
Comment (dsheets) [inline lib/profuse_7_8.ml]: `Getxattr` and `Setxattr` are missing the name extraction like in 7.23.

File: lib/profuse_7_8.ml [hunk 8]
```diff
@@ -792,18 +811,34 @@ module In = struct
...
+         | Removexattr name -> name
+         | Getlk t ->
+           Lk.describe t
+         | Setlk t ->
+           Lk.describe t
+         | Setlkw t ->
+           Lk.describe t
+         | Link (l,n) ->
+           let name = Link.name (to_voidp (addr l)) in
+           Printf.sprintf "name=%s %s" name n
```
Comment (dsheets) [inline lib/profuse_7_8.ml]: This is double printing the name and not printing the old node id.

File: lib/profuse_7_8.ml [hunk 9]
```diff
@@ -792,18 +811,34 @@ module In = struct
...
+         | Flush _r -> ""
+         | Fsyncdir _r -> ""
+         | Fsync _r -> ""
+         | Bmap _r -> ""
```
Comment (dsheets) [inline lib/profuse_7_8.ml]: This should print `FIXME` rather than empty string.

**Review by dsheets — CHANGES_REQUESTED** [review]
Just a few more changes -- sorry!

**Comment by yallop** [comment]
> Just a few more changes -- sorry!

I've addressed these in 78d8ba8.

**Review by dsheets — APPROVED** [review]

**Comment by dsheets** [comment]
LGTM. Thanks!

**Pull request merged**